Fix middleware tests

sowenjub opened the pull request:
Hey @excid3,
while this [pull request](https://github.com/gorails-screencasts/gorails-episode-234/pull/2/files) fixed integration tests, it broke the middleware tests.
I could have kept the `ActionDispatch::Request` introduced by the previous pull request but:
* introducing a Rails class in this piece of middleware felt wrong
* it would require a patch mixing `request.env.variable = ` and `request.env["VARIABLE"] = ` which felt inelegant.
* it seems unnecessary (I took the join idea from the way the rack webrick handlers handles it)

We could probably use `env["PATH_INFO"]` instead of the `[env["SCRIPT_NAME"], env["PATH_INFO"]].join` but in the off chance that another piece of middleware modifies `SCRIPT_NAME`, this seemed preferable.

I'm new to all of this, so I might be missing (or misunderstanding) something.

Also I don't know about REQUEST_PATH but I found that REQUEST_URI seems only provided by some application servers. I guess you hinted at that in the video, this left me wondering if there's a way to easily discover if a new variable must be added in the future.

sowenjub committed:
Fix middleware tests

excid3 commented:
@sowenjub Awesome, do you have a test to replicate the failing middleware test you mentioned?

excid3 commented:
And I think DHH or someone just shared a snippet of how they setup their middleware for Basecamp and it's along the same lines.

sowenjub commented:
> @sowenjub Awesome, do you have a test to replicate the failing middleware test you mentioned?

@excid3 Well yes, the `/middleware/account_middleware_test.rb` of the current master is fails without this patch.

> And I think DHH or someone just shared a snippet of how they setup their middleware for Basecamp and it's along the same lines.

Oh wow, if you manage to find it that would be interesting.

excid3 commented:
I'll see if I can find that tweet again. DHH tweets a lot though so we shall see. 😜

Also, I think this middleware test may not that great of a test. I wrote it fairly coupled to the original implementation I wrote, so not sure that it's that good. Was just reading through that again.

excid3 commented:
Found a little info on the `request_path` and `request_uri` 

https://stackoverflow.com/questions/1515108/whats-the-difference-between-request-envrequest-uri-and-request-envreques

sowenjub commented:
What's bad about it? I actually reverted to your original code, and simply added support for integration tests which have a nil `REQUEST_PATH` (which was what the other pull request was trying to address).

Regarding REQUEST_URI, I saw that as well as I was trying to understand things. Checking it was on my todo. Which I just did, and the current implementation does strip the query string.
So we just have to add env["QUERY_STRING"] I guess since it's part of the RFC it shouldn't go anywhere.

sowenjub commented:
I guess the we'd minimize the chances of messing with the REQUEST_URI by replacing
```
env["REQUEST_URI"]  = "/#{request_path}"
```
with
```
env["REQUEST_URI"]  = env["REQUEST_URI"]&.sub("/#{account_id}", "")
```
instead of trying to append the QUERY_STRING.

sowenjub commented:
So here's where I'm at.
lib/middlewares/account_middleware.rb
```
class AccountMiddleware
  def initialize(app)
    @app = app
  end

  def call(env)
    original_request_path = [env["SCRIPT_NAME"], env["PATH_INFO"]].join
    _, account_id, request_path = original_request_path.split("/", 3)

    if /\d+/.match?(account_id)
      if account = Account.find_by(id: account_id)
        Rails.logger.debug "[AccountMiddleware] Found account for id #{account_id}"
        Current.account = account
      else
        Rails.logger.debug "[AccountMiddleware] 404 - Found no account for id #{account_id}"
        return [302, { "Location" => "/" }, []]
      end

      env["SCRIPT_NAME"]  = "/#{account_id}"
      env["PATH_INFO"]    = "/#{request_path}"
      env["REQUEST_PATH"] = "/#{request_path}"
      env["REQUEST_URI"]  = env["REQUEST_URI"]&.sub("/#{account_id}", "")
    end

    @app.call(env)
  end
end
```
and test/middlewares/account_middleware_test.rb
I moved the REQUEST_URI in its own test since it might not always be there.
```
class AccountMiddlewareTest < ActiveSupport::TestCase
  setup do
    Current.account = nil
  end

  test "loads the account" do
    account      = accounts(:dundler_mifflin_scranton)
    account_path = "/#{account.id}"
    path         = "/teams"
    full_path    = account_path + path

    app        = ->(env) { [200, env, "app"] }
    middleware = AccountMiddleware.new(app)
    response   = Rack::MockRequest.new(middleware).get(full_path, "REQUEST_PATH" => full_path)

    assert_equal Current.account, account
    assert_equal account_path, response["SCRIPT_NAME"]
    assert_equal path, response["PATH_INFO"]
    assert_equal path, response["REQUEST_PATH"]
  end

  test "handles REQUEST_URI" do
    account      = accounts(:dundler_mifflin_scranton)
    account_path = "/#{account.id}"
    path         = "/teams"
    full_path    = account_path + path
    query_string = "?page=1"
    full_uri     = full_path + query_string

    app        = ->(env) { [200, env, "app"] }
    middleware = AccountMiddleware.new(app)
    response   = Rack::MockRequest.new(middleware).get(full_path, "REQUEST_PATH" => full_path, "REQUEST_URI" => full_uri)

    assert_equal Current.account, account
    assert_equal account_path, response["SCRIPT_NAME"]
    assert_equal path, response["PATH_INFO"]
    assert_equal path, response["REQUEST_PATH"]
    assert_equal path + query_string, response["REQUEST_URI"]
  end

  test "redirects if it fails to load the account" do
    account_path = "/12345"
    path         = "/teams"
    full_path    = account_path + path

    app        = ->(env) { [200, env, "app"] }
    middleware = AccountMiddleware.new(app)
    response   = Rack::MockRequest.new(middleware).get(full_path, "REQUEST_PATH" => full_path)

    assert_equal "/", response.headers["Location"]

    assert_nil Current.account
    assert_nil response["SCRIPT_NAME"]
    assert_nil response["PATH_INFO"]
    assert_nil response["REQUEST_PATH"]
    assert_nil response["REQUEST_URI"]
  end
end
```
